src/reader/warming.rs

```diff
@@ -0,0 +1,353 @@
+    ///
+    /// A background GC thread for [Warmer::garbage_collect] calls is uniquely created if there are active warmers.
+    pub fn new_searcher_generation(&self, searcher: &Searcher) -> crate::Result<()> {
+        self.0
+            .lock()
+            .unwrap()
+            .new_searcher_generation(searcher, &self.0)
+    }
+
+    #[cfg(test)]
+    fn gc_maybe(&self) -> bool {
+        self.0.lock().unwrap().gc_maybe()
+    }
+}
+
+struct WarmingStateInner {
+    num_warming_threads: usize,
+    warmers: Vec<Weak<dyn Warmer>>,
+    searcher_generations: HashMap<SearcherIndexGeneration, Vec<SearcherGenerationToken>>,
+    gc_thread: Option<JoinHandle<()>>,
+}
+
+impl WarmingStateInner {
+    /// Start tracking provided searcher as an exemplar of a new generation.
+    /// If there are active warmers, warm them with the provided searcher, and kick background GC thread if it has not yet been kicked.
+    /// Otherwise, prune state for dropped searcher generations inline.
+    fn new_searcher_generation(
+        &mut self,
+        searcher: &Searcher,
+        this: &Arc<Mutex<Self>>,
+    ) -> crate::Result<()> {
+        if let Some(tokens) = self
+            .searcher_generations
+            .get_mut(searcher.index_generation())
+        {
+            tokens.push(searcher.generation_token());
+        } else {
+            self.searcher_generations.insert(
+                searcher.index_generation().clone(),
+                vec![searcher.generation_token()],
+            );
+        }
+        let warmers = self.pruned_warmers();
+        // Avoid threads (warming as well as background GC) if there are no warmers
+        if warmers.is_empty() {
+            self.prune_searcher_generations();
+        } else {
+            self.start_gc_thread_maybe(this)?;
+            warming_executor(self.num_warming_threads.min(warmers.len()))?
+                .map(|warmer| warmer.warm(searcher), warmers.into_iter())?;
+        }
+        Ok(())
+    }
+
+    /// Attempt to upgrade the weak Warmer references, pruning those which cannot be upgraded.
+    /// Return the strong references.
+    fn pruned_warmers(&mut self) -> Vec<Arc<dyn Warmer>> {
+        let strong_warmers = self
+            .warmers
+            .iter()
+            .flat_map(|weak_warmer| weak_warmer.upgrade())
+            .collect::<Vec<_>>();
+        self.warmers = strong_warmers.iter().map(Arc::downgrade).collect();
+        strong_warmers
+    }
+
+    /// Prune dropped searcher generations from our state.
+    /// Return count of removed generations.
+    fn prune_searcher_generations(&mut self) -> usize {
+        for tokens in self.searcher_generations.values_mut() {
+            // This can be implemented neatly using Vec::drain_filter() when that is stable.
+            let mut i = 0;
+            while i < tokens.len() {
+                if !tokens[i].is_live() {
+                    tokens.remove(i);
```

Comment:
Performance does not matter but well.... `swap_remove` is more indicated here..

Reply:
ooh, makes sense!